### ISSUE TITLE
write(FH) on closed FH cuases segv

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2977]
Migrated from [rt.perl.org#4893](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=4893) (status was 'resolved')

Searchable as RT4893$

**Comment by p5pRT** [comment]
### From @mjdominus

#### Created by @mjdominus



The write\(\) itself doesn't cause the segv; it's just silently
ignored\.  Rather\, when you write past the end of the 'page'\, 
Perl tries to invoke the FH\_TOP format automatically\, and that cuases
the segv\.

I tried to fix it\, but it turned out to be more complicated than I
thought\.  Enclosed is a patch to t/op/misc\.t that installs a test case
for this bug\.

\-\-\- t/op/misc\.t    2000/12/10 19&#8203;:48&#8203;:23    1\.1
\+\+\+ t/op/misc\.t    2000/12/10 19&#8203;:51&#8203;:08    1\.2
@&#8203;@&#8203; \-545\,3 \+545\,23 @&#8203;@&#8203;
 lcfirst \- world
 uc \- WORLD
 lc \- world
\+\#\#\#\#\#\#\#\#
\+\# 20001210 mjd@&#8203;plover\.com
\+format REMITOUT\_TOP =
\+FOO
\+\.
\+
\+format REMITOUT =
\+BAR
\+\.
\+
\+\# This loop causes a segv in 5\.6\.0
\+for $lineno \(1\.\.61\) \{
\+   write REMITOUT;
\+\}
\+
\+print "It's OK\!";
\+EXPECT
\+It's OK\!
\+
\+  

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by mjd at Wed Sep 20 15:34:29 EDT 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.16, archname=i586-linux
    uname='linux plover 2.2.16 #97 fri jun 16 19:45:30 pdt 2000 i586 unknown '
    config_args='-Doptimize=-g -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-g', gccversion=egcs-2.91.66 19990314/Linux (egcs-1.1.2 release)
    cppflags='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccflags ='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt
    libc=/lib/libc-2.1.3.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/local/lib/perl5/5.6.0/i586-linux
    /usr/local/lib/perl5/5.6.0
    /usr/local/lib/perl5/site_perl/5.6.0/i586-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl/5.00557/i586-linux
    /usr/local/lib/perl5/site_perl/5.00557
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/mjd
    LANG (unset)
    LANGUAGE (unset)
    LC_ALL=POSIX
    LD_LIBRARY_PATH=/lib:/usr/lib:/usr/X11R6/lib
    LOGDIR (unset)
    PATH=/home/mjd/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11/bin:/usr/games:/sbin:/usr/sbin:/usr/local/bin/X11:/usr/local/bin/mh:/data/mysql/bin:/usr/local/bin/pbm:/usr/local/bin/ezmlm:/home/mjd/TPI/bin:/usr/local/mysql/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

\<mjd@&#8203;plover\.com> writes&#8203;:
>This is a bug report for perl from mjd@&#8203;plover\.plover\.com\,
>generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.
>
>
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>\[Please enter your report here\]
>
>The write\(\) itself doesn't cause the segv; it's just silently
>ignored\.  Rather\, when you write past the end of the 'page'\, 
>Perl tries to invoke the FH\_TOP format automatically\, and that cuases
>the segv\.

Having just visited the format code to make it utf8 aware I am not 
in the least surprised by the bug\.

It is just possible that it got fixed as a result of what I did\,
if not I am more or less up to speed\.

Thanks for the test though\.

>
>I tried to fix it\, but it turned out to be more complicated than I
>thought\.  Enclosed is a patch to t/op/misc\.t that installs a test case
>for this bug\.
>
>\-\-\- t/op/misc\.t    2000/12/10 19&#8203;:48&#8203;:23    1\.1
>\+\+\+ t/op/misc\.t    2000/12/10 19&#8203;:51&#8203;:08    1\.2
>@&#8203;@&#8203; \-545\,3 \+545\,23 @&#8203;@&#8203;
> lcfirst \- world
> uc \- WORLD
> lc \- world
>\+\#\#\#\#\#\#\#\#
>\+\# 20001210 mjd@&#8203;plover\.com
>\+format REMITOUT\_TOP =
>\+FOO
>\+\.
>\+
>\+format REMITOUT =
>\+BAR
>\+\.
>\+
>\+\# This loop causes a segv in 5\.6\.0
>\+for $lineno \(1\.\.61\) \{
>\+   write REMITOUT;
>\+\}
>\+
>\+print "It's OK\!";
>\+EXPECT
>\+It's OK\!
>\+
>\+  
>
>
>
>\[Please do not change anything below this line\]
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>\-\-\-
>Flags&#8203;:
>    category=core
>    severity=medium
>\-\-\-
>Site configuration information for perl v5\.6\.0&#8203;:
>
>Configured by mjd at Wed Sep 20 15&#8203;:34&#8203;:29 EDT 2000\.
>
>Summary of my perl5 \(revision 5\.0 version 6 subversion 0\) configuration&#8203;:
>  Platform&#8203;:
>    osname=linux\, osvers=2\.2\.16\, archname=i586\-linux
>    uname='linux plover 2\.2\.16 \#97 fri jun 16 19&#8203;:45&#8203;:30 pdt 2000 i586 unknown '
>    config\_args='\-Doptimize=\-g \-des'
>    hint=recommended\, useposix=true\, d\_sigaction=define
>    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
>    useperlio=undef d\_sfio=undef uselargefiles=define 
>    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
>  Compiler&#8203;:
>    cc='cc'\, optimize='\-g'\, gccversion=egcs\-2\.91\.66 19990314/Linux \(egcs\-1\.1\.2 release\)
>    cppflags='\-DDEBUGGING \-fno\-strict\-aliasing \-I/usr/local/include'
>    ccflags ='\-DDEBUGGING \-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'
>    stdchar='char'\, d\_stdstdio=define\, usevfork=false
>    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
>    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
>    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
>    alignbytes=4\, usemymalloc=n\, prototype=define
>  Linker and Libraries&#8203;:
>    ld='cc'\, ldflags =' \-L/usr/local/lib'
>    libpth=/usr/local/lib /lib /usr/lib
>    libs=\-lnsl \-lndbm \-lgdbm \-ldb \-ldl \-lm \-lc \-lposix \-lcrypt
>    libc=/lib/libc\-2\.1\.3\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
>  Dynamic Linking&#8203;:
>    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
>    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'
>
>Locally applied patches&#8203;:
>    
>
>\-\-\-
>@&#8203;INC for perl v5\.6\.0&#8203;:
>    /usr/local/lib/perl5/5\.6\.0/i586\-linux
>    /usr/local/lib/perl5/5\.6\.0
>    /usr/local/lib/perl5/site\_perl/5\.6\.0/i586\-linux
>    /usr/local/lib/perl5/site\_perl/5\.6\.0
>    /usr/local/lib/perl5/site\_perl/5\.00557/i586\-linux
>    /usr/local/lib/perl5/site\_perl/5\.00557
>    /usr/local/lib/perl5/site\_perl
>    \.
>
>\-\-\-
>Environment for perl v5\.6\.0&#8203;:
>    HOME=/home/mjd
>    LANG \(unset\)
>    LANGUAGE \(unset\)
>    LC\_ALL=POSIX
>    LD\_LIBRARY\_PATH=/lib&#8203;:/usr/lib&#8203;:/usr/X11R6/lib
>    LOGDIR \(unset\)
>    PATH=/home/mjd/bin&#8203;:/usr/local/bin&#8203;:/bin&#8203;:/usr/bin&#8203;:/usr/X11/bin&#8203;:/usr/games&#8203;:/sbin&#8203;:/usr/sbin&#8203;:/usr/local/bin/X11&#8203;:/usr/local/bin/mh&#8203;:/data/mysql/bin&#8203;:/usr/local/bin/pbm&#8203;:/usr/local/bin/ezmlm&#8203;:/home/mjd/TPI/bin&#8203;:/usr/local/mysql/bin
>    PERL\_BADLANG \(unset\)
>    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @mjdominus


> It is just possible that it got fixed as a result of what I did\,
> if not I am more or less up to speed\.

I really doubt it\.  It has to do with perl getting a NULL stream
object and then passing the null pointer to PerlIO\_write\.  The dump
occurs at pp\_sys\.c&#8203;:1247&#8203;:

&nbsp;    if \(IoLINES\_LEFT\(io\) >= 0 && IoPAGE\(io\) > 0\) 
&nbsp;            PerlIO\_write\(ofp\, SvPVX\(PL\_formfeed\)\, SvCUR\(PL\_formfeed\)\);

I tried guarding this with a test for null 'ofp'&#8203;:

&nbsp;    if \(IoLINES\_LEFT\(io\) >= 0 && IoPAGE\(io\) > 0\)
&nbsp;            if \(ofp\) 
&nbsp;              PerlIO\_write\(ofp\, SvPVX\(PL\_formfeed\)\, SvCUR\(PL\_formfeed\)\);
&nbsp;        else if \(ckWARN\(WARN\_CLOSED\)\)
&nbsp;        report\_closed\_fh\(gv\, io\, "write"\, "filehandle"\);


but that wasn't enough; there's something else going on also\.  I
investigated further\, but wasn't able to figure it out\.

Anyway\, I don't think it was your fault\, and just to be sure\, I
checked it with 5\.000\, which aldso exhibits the problem\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Mark\-Jason Dominus \<mjd@&#8203;plover\.com> writes&#8203;:
>> It is just possible that it got fixed as a result of what I did\,
>> if not I am more or less up to speed\.
>
>I really doubt it\.  It has to do with perl getting a NULL stream
>object and then passing the null pointer to PerlIO\_write\.  

I guessed as much \- which is why I think it may be fixed ;\-\)
With the utf8 patches there are \_NO\_ calls to PerlIO\_write\(\) in the 
format code \- they have all become calls to do\_print\(\)\, so \(depending
on where check for NULL fp is\) the problem has at least moved\.\.\.

>
>Anyway\, I don't think it was your fault\, and just to be sure\, I
>checked it with 5\.000\, which aldso exhibits the problem\.

I am sure it wasn't my fault \- I have never been near that code before \.\.\. :\-\)

**Comment by p5pRT** [comment]
### From @mjdominus


> I guessed as much \- which is why I think it may be fixed ;\-\)

Sorry; I somehow read 'broken' where you had 'fixed'\.  The rest of my
response makes no sense without this\.

**Comment by p5pRT** [comment]
### From @jhi

On Mon\, Dec 11\, 2000 at 10&#8203;:39&#8203;:35AM \-0500\, Mark\-Jason Dominus wrote&#8203;:
> 
> > I guessed as much \- which is why I think it may be fixed ;\-\)
> 
> Sorry; I somehow read 'broken' where you had 'fixed'\.  The rest of my
> response makes no sense without this\.

I tried it out and it doesn't seem to be fixed yet\.\.\.

**Comment by p5pRT** [comment]
### From @cpansprout

The crash was fixed in 7ef822cddfe\, but ticket \#17099 says there is
still a memory leak\.

\-\- 

Father Chrysostomos
